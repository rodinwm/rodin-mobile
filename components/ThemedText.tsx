import {Text, type TextProps} from 'react-native';
import {useThemeColor} from '@/hooks/useThemeColor';
import {FontHelper} from "@/utils/helpers/fontHelper";
import {FontWeightEnum} from "@/utils/enums";

export type ThemedTextProps = TextProps & {
    lightColor?: string;
    darkColor?: string;
    type?: 'default' | 'title' | 'defaultSemiBold' | 'subtitle' | 'link';
};

export function ThemedText({
                               lightColor,
                               darkColor,
                               type = 'default',
                               ...rest
                           }: ThemedTextProps
) {
    const color = useThemeColor({light: lightColor, dark: darkColor}, 'text');
    const classNames = {
        default: "text-base",
        defaultSemiBold: "text-base",
        title: "text-3xl",
        subtitle: "text-xl",
        link: "text-base text-blue-500",
    }
    const fontWeights = {
        default: FontWeightEnum.Light,
        defaultSemiBold: FontWeightEnum.SemiBold,
        title: FontWeightEnum.ExtraBold,
        subtitle: FontWeightEnum.Bold,
        link: FontWeightEnum.Medium,
    }

    return (
        <Text
            className={classNames[type]}
            style={{
                color,
                fontFamily: FontHelper.getMainFont(fontWeights[type]),
            }}
            {...rest}
        />
    );
}