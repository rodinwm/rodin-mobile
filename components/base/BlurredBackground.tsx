import {BlurView} from "expo-blur";
import React from "react";

export function BlurredBackground() {
    return (
        <BlurView
            tint="dark"
            className={'flex-1'}
            intensity={30}
            style={{position: 'absolute', top: 0, left: 0, right: 0, bottom: 0}}
        />
    );
};

