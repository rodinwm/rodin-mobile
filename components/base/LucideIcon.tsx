import {icons} from 'lucide-react-native';
import {Colors} from "@/utils/colors";
import {useColorScheme} from "@/utils/hooks/useColorScheme";
import {StyleProp, ViewStyle} from "react-native";

type Props = {
    name: keyof typeof icons,
    color?: string,
    size?: number,
    className?: string,
    inverseColor?: boolean;
    strokeWidth?: number;
    style?: StyleProp<ViewStyle>;
}

export default function LucideIcon(props: Props) {
    const LucideIcon = icons[props.name];

    let colorScheme = useColorScheme();
    if (props.inverseColor && colorScheme) {
        colorScheme = (colorScheme == 'light') ? 'dark' : 'light';
    }

    return <LucideIcon
        color={props.color ?? Colors.foreground[colorScheme ?? 'light']}
        className={`flex-shrink-0 ${props.className}`}
        size={props.size}
        style={props.style}
        strokeWidth={props.strokeWidth}
        //strokeWidth={2.5}
    />;
};

