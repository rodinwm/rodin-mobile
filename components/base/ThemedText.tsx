import {Text, type TextProps} from 'react-native';
import {FontHelper} from "@/utils/helpers/fontHelper";
import {FontWeightEnum} from "@/utils/enums";

export type ThemedTextProps = TextProps & {
    type?: 'mini' | 'miniBold' | 'miniExtraBold' | 'default' | 'bigTitle' | 'title' | 'h1' | 'defaultSemiBold' | 'defaultExtraBold' | 'subtitle' | 'link' | 'logo';
    filled?: boolean;
    inverseColor?: boolean;
    numberOfLines?: number;
};

export function ThemedText({
                               className,
                               numberOfLines,
                               type = 'default',
                               filled = true,
                               inverseColor = false,
                               ...otherProps
                           }: ThemedTextProps
) {
    const style = {
        mini: {
            fontSize: "text-xs",
            fontWeight: FontWeightEnum.Light,
        },
        miniBold: {
            fontSize: "text-xs",
            fontWeight: FontWeightEnum.Bold,
        },
        miniExtraBold: {
            fontSize: "text-xs",
            fontWeight: FontWeightEnum.ExtraBold,
        },
        link: {
            fontSize: "text-lg text-blue-500",
            fontWeight: FontWeightEnum.Medium,
        },
        default: {
            fontSize: "text-lg",
            fontWeight: FontWeightEnum.Light,
        },
        defaultSemiBold: {
            fontSize: "text-lg",
            fontWeight: FontWeightEnum.SemiBold,
        },
        defaultExtraBold: {
            fontSize: "text-lg",
            fontWeight: FontWeightEnum.ExtraBold,
        },
        subtitle: {
            fontSize: "text-xl",
            fontWeight: FontWeightEnum.Bold,
        },
        h1: {
            fontSize: "text-2xl",
            fontWeight: FontWeightEnum.SemiBold,
        },
        title: {
            fontSize: "text-3xl",
            fontWeight: FontWeightEnum.ExtraBold,
        },
        bigTitle: {
            fontSize: "text-5xl",
            fontWeight: FontWeightEnum.ExtraBold,
        },
        logo: {
            fontSize: "text-8xl",
            fontWeight: FontWeightEnum.ExtraBold,
        },
    };

    const classNames: string[] = [
        style[type].fontSize,
        filled ?
            !inverseColor ?
                'text-foreground-light dark:text-foreground-dark'
                : 'text-foreground-dark dark:text-foreground-light'
            : '',
        type === "link" ? 'underline underline-offset-8' : '',
        className ?? ''
    ];

    return (
        <Text
            className={classNames.join(' ')}
            style={{
                fontFamily: type === 'bigTitle' ?
                    FontHelper.getBigTitleFontStatic() : FontHelper.getMainFontStatic(style[type].fontWeight)
            }}
            numberOfLines={numberOfLines}
            ellipsizeMode="tail"
            {...otherProps}
        />
    );
}