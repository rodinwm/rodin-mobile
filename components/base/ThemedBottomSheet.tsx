import BottomSheet, {BottomSheetView} from "@gorhom/bottom-sheet";
import React, {ReactNode, useEffect, useRef} from "react";
import {ThemedView} from "@/components/base/ThemedView";
import {Pressable} from "react-native";
import {BlurView} from "expo-blur";

export type ThemedBottomSheetProps = {
    isOpen: boolean;
    children?: ReactNode;
    onClose?: () => void;
};

export default function ThemedBottomSheet(props: ThemedBottomSheetProps) {
    const ref = useRef<BottomSheet>(null);
    const margin = 16;

    useEffect(() => {
        if (props.isOpen) {
            ref.current?.expand();
        } else {
            ref.current?.close();
        }
    }, [props.isOpen]);

    return (
        <>
            {/* Flou d'arri√®re-plan */}
            <Pressable
                className={props.isOpen ? '' : 'hidden'}
                onPress={props.onClose}
                style={{
                    position: 'absolute',
                    top: 0,
                    left: 0,
                    right: 0,
                    bottom: 0,
                }}
            >
                <BlurView intensity={50} tint="dark" style={{flex: 1}}/>
            </Pressable>

            {/* Bottom sheet */}
            <BottomSheet
                ref={ref}
                index={-1}
                detached={true}
                enablePanDownToClose={true}
                bottomInset={margin}
                onClose={props.onClose}
                style={{marginHorizontal: margin}}
                handleComponent={() => (
                    <ThemedView
                        paddingStyle={"mini"}
                        className={'w-full flex items-center justify-center'}
                    >
                        <ThemedView
                            radiusStyle={"full"}
                            className={'w-10 h-2'}
                            fillStyle={"opacity-15"}
                        />
                    </ThemedView>
                )}
                backgroundComponent={({style}) => (
                    <ThemedView
                        style={[style, {borderRadius: 36}]}
                        fillStyle={"default"}
                        radiusStyle={"big"}
                        outlined={true}
                    />
                )}
            >
                <BottomSheetView className={'p-6 flex flex-col gap-3'}>
                    {props.children}
                </BottomSheetView>
            </BottomSheet>
        </>
    );

}