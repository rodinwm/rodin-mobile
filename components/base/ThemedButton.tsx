import LucideIcon from "@/components/base/LucideIcon";
import {type ButtonProps, TouchableOpacity} from "react-native";
import React from "react";
import {ThemedText} from "@/components/base/ThemedText";
import {icons} from "lucide-react-native";

export type ThemedButtonProps = ButtonProps & {
    icon?: {
        name: keyof typeof icons;
        size?: number;
    };
    showTitle?: boolean;
    fullWidth?: boolean;
    fullHeight?: boolean;
    miniText?: boolean;
    className?: string;
    radiusStyle?: "default" | "full" | "left-only" | "right-only" | "none";
    type?: 'default' | 'outlined' | 'danger' | 'success' | 'link';
};

export function ThemedButton({
                                 icon,
                                 title,
                                 className,
                                 fullWidth,
                                 fullHeight,
                                 miniText,
                                 showTitle = true,
                                 radiusStyle = "default",
                                 type = "default",
                                 ...otherProps
                             }: ThemedButtonProps
) {
    const classNames: string[] = [
        'flex flex-row gap-2 items-center justify-center border p-3',
        showTitle ? 'px-6 py-3' : '',
        fullWidth ? 'w-full' : '',
        fullHeight ? 'h-full' : '',
        radiusStyle === "default" ?
            'rounded-2xl' : radiusStyle === "full" ?
                'rounded-full' : radiusStyle === "left-only" ?
                    'rounded-l-2xl' : radiusStyle === "right-only" ?
                        'rounded-r-2xl' : '',
        type === "default" ?
            'bg-background-dark dark:bg-background-light border-foreground-light dark:border-foreground-dark' : type === "outlined" ?
                'bg-background-dark/10 dark:bg-background-light/10 border-foreground-light/20 dark:border-foreground-dark/20' : type === "danger" ?
                    'bg-red-700 border-red-700' : type === "success" ?
                        'bg-emerald-700 border-emerald-700' : type === "link" ?
                            'px-0 py-0 opacity-50 border-transparent' : '',
        className ?? '',
    ];

    return (
        <TouchableOpacity
            className={classNames.join(' ')}
            {...otherProps}
        >
            {icon && (
                <LucideIcon
                    size={icon.size}
                    name={icon.name}
                    inverseColor={type === "default"}
                />
            )}

            {showTitle && (
                <ThemedText
                    type={
                        miniText ?
                            "miniBold" : type === "link" ?
                                'link' : "defaultSemiBold"
                    }
                    filled={type !== "danger"}
                    className={type === "danger" ? 'text-foreground-dark' : ''}
                    inverseColor={type === "default"}
                >
                    {title}
                </ThemedText>
            )}
        </TouchableOpacity>
    );
}
