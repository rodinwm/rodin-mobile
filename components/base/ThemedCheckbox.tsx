import {Colors} from "@/utils/colors";
import {useColorScheme} from "@/utils/hooks/useColorScheme";
import {LucideIcon} from "@/components/base/LucideIcon";
import BouncyCheckbox, {BouncyCheckboxProps} from "react-native-bouncy-checkbox";
import React from "react";

export type ThemedCheckboxProps = BouncyCheckboxProps & {}

export function ThemedCheckbox({
                                   ...otherProps
                               }: ThemedCheckboxProps
) {
    const colorScheme = useColorScheme() ?? 'light';

    return (
        <BouncyCheckbox
            iconComponent={(
                <LucideIcon name={"Check"} size={14} inverseColor={true}/>
            )}
            fillColor={Colors.foreground[colorScheme]}
            //unFillColor={Colors.foreground[colorScheme] + "33"} // FIXME: find how to hide checkkk icon when checkbox is not checked
            disableText={true}
            {...otherProps}
        />
    );
};

