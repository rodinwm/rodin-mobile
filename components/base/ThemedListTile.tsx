import {LucideIcon} from "@/components/base/LucideIcon";
import {ThemedText} from "@/components/base/ThemedText";
import {ThemedView} from "@/components/base/ThemedView";
import {type ButtonProps, TouchableOpacity} from "react-native";
import {icons} from "lucide-react-native";
import {isValidElement, ReactNode} from "react";
import {UIHelper} from "@/utils/helpers/UIHelper";
import {Colors} from "@/utils/colors";
import {useColorScheme} from "@/utils/hooks/useColorScheme";

export type ThemedListTileProps = ButtonProps & {
    subtitle?: string;
    icon?: keyof typeof icons | ReactNode;
    suffixIcon?: keyof typeof icons | ReactNode | null;
    fillStyle?: "default" | "opacity-15" | "opacity-50" | "warning" | "inversed" | "none";
    hasPadding?: boolean;
    hapticOnPress?: boolean;
};

export function ThemedListTile({
                                   icon,
                                   title,
                                   subtitle,
                                   hasPadding,
                                   suffixIcon = 'ChevronRight',
                                   fillStyle = "none",
                                   hapticOnPress = true,
                                   onPress,
                                   ...otherProps
                               }: ThemedListTileProps
) {
    const colorScheme = useColorScheme() ?? 'light';

    return (
        <TouchableOpacity
            onPress={onPress !== undefined ? (event) => {
                if (hapticOnPress) UIHelper.hapticImpact()
                onPress(event);
            } : undefined}
            {...otherProps}
        >
            <ThemedView
                fillStyle={otherProps.disabled ? "opacity-50" : fillStyle}
                radiusStyle={hasPadding ? "default" : "none"}
                paddingStyle={hasPadding ? "small" : "none"}
                className={'w-full flex flex-row justify-between items-center gap-2'}
            >
                <ThemedView className={'h-fit flex flex-row justify-center items-center gap-4 flex-1'}>
                    {icon !== undefined && icon !== null ?
                        isValidElement(icon) ? (icon) : (
                            fillStyle === "none" ? (
                                <LucideIcon name={icon as keyof typeof icons}/>
                            ) : (
                                <ThemedView
                                    fillStyle={fillStyle !== "inversed" ? "inversed" : 'default'}
                                    radiusStyle={'full'}
                                    className={"p-2"}
                                    borderStyle={"default"}
                                >
                                    <LucideIcon
                                        name={icon as keyof typeof icons}
                                        color={otherProps.disabled ? Colors.foreground[colorScheme] + '66' : undefined}
                                        inverseColor={fillStyle !== "inversed"}
                                    />
                                </ThemedView>
                            ))
                        : null}

                    <ThemedView className={'flex-1 flex flex-col justify-center'}>
                        <ThemedText
                            type={'defaultSemiBold'}
                            numberOfLines={2}
                            inverseColor={fillStyle === "inversed"}
                        >
                            {title}
                        </ThemedText>
                        {subtitle && (
                            <ThemedText
                                type={'small'}
                                inverseColor={fillStyle === "inversed"}
                                className={"opacity-50"}
                            >
                                {subtitle}
                            </ThemedText>
                        )}
                    </ThemedView>
                </ThemedView>

                {suffixIcon !== undefined && suffixIcon !== null ? (
                    isValidElement(suffixIcon) ? (suffixIcon) : (
                        <LucideIcon
                            size={20}
                            name={suffixIcon as keyof typeof icons}
                            inverseColor={fillStyle === "inversed"}
                        />
                    )
                ) : null}

            </ThemedView>
        </TouchableOpacity>


    );
};

