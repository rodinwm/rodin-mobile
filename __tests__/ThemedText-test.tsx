import * as React from 'react';
import renderer from 'react-test-renderer';

import {ThemedText} from '@/components/base/ThemedText';

it(`renders correctly`, () => {
    const tree = renderer.create(<ThemedText>test</ThemedText>).toJSON();

    expect(tree).toMatchSnapshot();
});
