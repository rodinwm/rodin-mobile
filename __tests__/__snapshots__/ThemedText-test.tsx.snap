// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<Text
  className="text-lg text-foreground-light dark:text-foreground-dark "
  ellipsizeMode="tail"
  style={
    {
      "fontFamily": "FunnelDisplay-Light",
    }
  }
>
  test
</Text>
`;
